compiler/rustc_data_structures/src/profiling.rs

```diff
@@ -443,6 +509,47 @@ impl SelfProfilerRef {
     }
 }
 
+/// A helper for recording costly arguments to self-profiling events. Used with
+/// `SelfProfilerRef::generic_activity_with_arg_recorder`.
+pub struct EventArgRecorder<'p> {
+    /// The resulting event id we want to use within the generic activity's `TimingGuard`.
+    event_id: &'p mut EventId,
+
+    /// The interned label to be recorded in the generic activity event.
+    event_label: StringId,
+
+    /// The `SelfProfiler` used to intern the event arguments that users will ask to record.
+    profiler: &'p SelfProfiler,
+
+    /// The `measureme` builder creating the event id we need, by using the interned event label and
+    /// interned arguments.
+    builder: EventIdBuilder<'p>,
+}
+
+impl EventArgRecorder<'_> {
+    /// Records a single argument within the current generic activity being profiled.
+    ///
+    /// Note: when self-profiling with costly event arguments, either of the `record` methods
+    /// *needs* to be called. A panic will be triggered if that doesn't happen.
+    pub fn record_arg<A>(self, event_arg: A)
+    where
+        A: Borrow<str> + Into<String>,
+    {
+        let event_arg = self.profiler.get_or_alloc_cached_string(event_arg);
+        *self.event_id = self.builder.from_label_and_arg(self.event_label, event_arg);
+    }
```

Comment:
Just a thought, but it might be convenient if this didn't immediately allocate the `EventId` but instead buffered all the calls to `record_arg` in an internal vec (or smallvec) and then in `generic_activity_with_arg_recorder`, it could call `builder.from_label_and_arg` there. We could then potentially get rid of `record_args` in favor of calling `record_arg` multiple times. 

(Not saying we *should* do this, just an idea 🙂)

Reply:
We could totally do that. I was thinking multiple calls to record a single argument could be maybe too verbose for users of the self-profiler, but there are currently no multiple-argument activities that I could see anyways.

and if the slight ergonomic boost becomes needed one day, it’ll be easy to add record_args back in.

I’ll do that then, thank you.

Reply:
I've pushed this change, and ported the one `generic_activity_with_args` call I could find to multiple `record_arg` calls, and [it's not that bad](https://github.com/rust-lang/rust/pull/95689/commits/1906b7e967f98e8c735054af61906e86cfc80ac7).